debian/mongoperf.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOPERF" "1" "January 30, 2015" "3.0" "mongodb-manual"
.SH NAME
mongoperf \- MongoDB Performance Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBmongoperf\fP is a utility to check disk I/O performance
independently of MongoDB.
.sp
It times tests of random disk I/O and presents the results. You can
use \fBmongoperf\fP for any case apart from MongoDB. The
\fBmmf\fP \fBtrue\fP mode is completely generic. In
that mode it is somewhat analogous to tools such as \fI\%bonnie++\fP (albeit mongoperf is
simpler).
.sp
Specify options to \fBmongoperf\fP using a JavaScript document.
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fI\%bonnie\fP
.IP \(bu 2
\fI\%bonnie++\fP
.IP \(bu 2
\fI\%Output from an example run\fP
.IP \(bu 2
\fI\%Checking Disk Performance with the mongoperf Utility\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH OPTIONS
.INDENT 0.0
.TP
.B mongoperf
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help, \-h
Returns information on the options and use of \fBmongoperf\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B <jsonconfig>
\fBmongoperf\fP accepts configuration options in the form of a
file that holds a \fIJSON\fP document. You must stream the
content of this file into \fBmongoperf\fP, as in the following
operation:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongoperf < config
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this example \fBconfig\fP is the name of a file that holds a JSON
document that resembles the following example:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
{
  nThreads:<n>,
  fileSizeMB:<n>,
  sleepMicros:<n>,
  mmf:<bool>,
  r:<bool>,
  w:<bool>,
  recSizeKB:<n>,
  syncDelay:<n>
}
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See the \fI\%Configuration Fields\fP section for documentation of each
of these fields.
.UNINDENT
.SH CONFIGURATION FIELDS
.INDENT 0.0
.TP
.B mongoperf.nThreads
\fIType:\fP Integer.
.sp
\fIDefault:\fP 1
.sp
Defines the number of threads \fBmongoperf\fP will use in the
test. To saturate your system\(aqs storage system you will need
multiple threads. Consider setting \fBnThreads\fP to \fB16\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.fileSizeMB
\fIType:\fP Integer.
.sp
\fIDefault:\fP 1 megabyte (i.e. 1024\s-2\u2\d\s0 bytes)
.sp
Test file size.
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.sleepMicros
\fIType:\fP Integer.
.sp
\fIDefault:\fP 0
.sp
\fBmongoperf\fP will pause for the number of specified
\fBsleepMicros\fP divided by the
\fBnThreads\fP between each operation.
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.mmf
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBmmf\fP to \fBtrue\fP to use memory mapped
files for the tests.
.sp
Generally:
.INDENT 7.0
.IP \(bu 2
when \fBmmf\fP is \fBfalse\fP, \fBmongoperf\fP
tests direct, physical, I/O, without caching. Use a large file
size to test heavy random I/O load and to avoid I/O coalescing.
.IP \(bu 2
when \fBmmf\fP is \fBtrue\fP, \fBmongoperf\fP
runs tests of the caching system, and can use normal file system
cache. Use \fBmmf\fP in this mode to test file system cache
behavior with memory mapped files.
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.r
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBr\fP to \fBtrue\fP to perform reads as part of
the tests.
.sp
Either \fBr\fP or \fBw\fP must be \fBtrue\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.w
\fIType:\fP Boolean.
.sp
\fIDefault:\fP \fBfalse\fP
.sp
Set \fBw\fP to \fBtrue\fP to perform writes as part of
the tests.
.sp
Either \fBr\fP or \fBw\fP must be \fBtrue\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.recSizeKB
New in version 2.4.

.sp
\fIType:\fP Integer.
.sp
\fIDefault:\fP 4 kb
.sp
The size of each write operation.
.UNINDENT
.INDENT 0.0
.TP
.B mongoperf.syncDelay
\fIType:\fP Integer.
.sp
\fIDefault:\fP 0
.sp
Seconds between disk flushes. \fBmongoperf.syncDelay\fP is
similar to \fI\-\-syncdelay\fP for \fBmongod\fP\&.
.sp
The \fBsyncDelay\fP controls how frequently
\fBmongoperf\fP performs an asynchronous disk flush of the memory
mapped file used for testing. By default, \fBmongod\fP
performs this operation every 60 seconds. Use
\fBsyncDelay\fP to test basic system performance of
this type of operation.
.sp
Only use \fBsyncDelay\fP in conjunction with
\fBmmf\fP set to \fBtrue\fP\&.
.sp
The default value of \fB0\fP disables this.
.UNINDENT
.SH USE
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongoperf < jsonconfigfile
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Replace \fBjsonconfigfile\fP with the path to the \fBmongoperf\fP
configuration. You may also invoke \fBmongoperf\fP in the
following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
echo "{nThreads:16,fileSizeMB:10000,r:true,w:true}" | mongoperf
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this operation:
.INDENT 0.0
.IP \(bu 2
\fBmongoperf\fP tests direct physical random read and write io\(aqs, using
16 concurrent reader threads.
.IP \(bu 2
\fBmongoperf\fP  uses a 10 gigabyte test file.
.UNINDENT
.sp
Consider using \fBiostat\fP, as invoked in the following example to
monitor I/O performance during the test.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
iostat \-xtm 1
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
MongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


NEWS
WiredTiger release 2.5.0, 2014-12-24
------------------------------------

The WiredTiger 2.5.0 release contains significant new features, API changes
and many bug fixes.

Now that WiredTiger is part of MongoDB, we are tracking issues related to
MongoDB usage of WiredTiger in the MongoDB JIRA system.  Some entries in
the changelog now reference JIRA tickets that can be found at:

  http://jira.mongodb.org

New features and API changes:

* Add support for storing large values on-page in a btree rather than in
  an overflow item.  This is useful for workloads that want to keep large
  items in cache - since WiredTiger overflow items are never cached.
  It is configured via new `leaf_value_max` configuration setting.  This
  enhancement led to the deprecation of internal_item_max and leaf_item_max
  configuration settings, see upgrading documentation for further
  information. [#1282]

* Add support for compressing log files.  When configured each compressable
  log record will be compressed.  This is configure with the
  `log=(compressor=X)` configuration setting.  See upgrading documentation
  for further information. [#1359]

* No longer return EBUSY when opening a bulk cursor, verifying or salvaging
  a database if a checkpoint is currently running.  This allows
  applications to do these exclusive operations without shutting down the
  checkpoint server thread.  [#1397] [#1404] SERVER-16236 SERVER-16457

* Add support for immutable indexes. [#1344]

* Added several new statistics, improved accuracy for some statistics
  tracking and simplified mechanism for querying a particular statistic.
  [#1505]

* Have the eviction server write out unnecessary pages prior to the cache
  reaching the configured eviction trigger size.  This can reduce the
  amount of eviction application threads do when configured with a large
  cache.

* Several enhancements to managing how long we keep files open

* Revert a change in the 2.4.1 release that caused the WT_ROLLBACK (and
  deprecated WT_DEADLOCK) error return to map to different numeric values.
  Applications should ensure they are compiling against the same version
  of the wiredtiger.h header file as the library they link against,
  otherwise odd behavior will be experienced.

* Support setting configuration strings to "none" as being equivalent to an
  empty string in most cases. [#1417]

* Enhance the hot backup implementation to allow recovery to be run between
  incremental backups. [#1183]

Other significant changes:

* Improve performance of cursor open when there are many tables in a database.
  [#1391] [#1443]

* Reduce the impact checkpoints have on concurrent operations. This was
  done by changing how we lock tables. [#1391] [#1392]

* Improve performance when scanning a table that has many deleted items.
  SERVER-16247

* Fix a bug in checkpoint, where the metadata (turtle) file wasn't being
  synced on checkpoint. [#1383]

* Fix a bug where WiredTiger could accumulate memory during page splits and
  never free it.  SERVER-16546

* Many enhancements and bug fixes for Windows.

* Fix a bug where a custom extractor terminate was being called twice.
  [#1503]

* Fix a bug where a race between closing a handle and checkpointing could
  lead to errors. [#1495] [#1497]

* Validate the block header checksum before we clear it - if the checksum
  field had been corrupted, we didn't notice.  SERVER-16457

* Fix a bug in write conflict detection.  Cursors configured with
  no-overwrite could sometimes not see update conflicts for deleted
  records.  SERVER-16351

* Several bug fixes and performance improvements in LSM including:
 - Fix a bug in LSM search_near, where it return a deleted item.
   BF-694, BF-700
 - Improve background maintenance operation so that it doesn't fill the
   cache.
 - Fix a bug that could lead to updates being written into old chunks.
   [#1432] [#1418]
 - Fix a bug in background merge that could skip updates.  SERVER-16123
 - Add support for custom collators in LSM trees. [#1361]

* Fix a bug when maintaining the cache, that could cause checkpoints to
  skip writing an update that should have been included. [#1419] SERVER-16336

* Fix a bug in WT_SESSION::drop, where failures generated error output even
  when force was specified. [#1436]

* Fix a bug in WiredTiger integer packing code when figuring out how much
  space is required for a value (it can shrink as numbers grow). SERVER-16118

* Several enhancements to the wtstats.py tool that generates graphs from
  standard WiredTiger statistics logs. [#1365]

* Fix a bug on OS X where fsync isn't sufficient to flush a file, use
  fcntl(F_FULLFSYNC) instead.

* Work around a bug in clang 3.5.0 compare and swap primitive. The
  __sync_bool_compare_and_swap version of the API in clang produces bad
  code for us with -O3 optimization enabled.


WiredTiger release 2.4.1, 2014-11-06
------------------------------------

The WiredTiger 2.4.1 release contains several new features, many bug fixes
and performance enhancements.

New features and API changes:

* Add new custom extractor functionality to WiredTiger indexes. Allowing an
  application to define mutated and/or multiple keys for indexes. [#1199]
 
* Add a new WT_SESSION::transaction_pinned_range method that allows users
  to identify when a session is keeping a transaction ID pinned for a long
  time. [#1314]

* Enhance statistics output so that keys are more clearly categorized. [#1313]

* Rename WT_DEADLOCK error return to WT_ROLLBACK. WiredTiger uses the return
  in cases other than traditional application deadlock. The old value is
  retained as an alias to maintain backward compatability. [#1204]

* Increase the maximum configurable cache size to 100GB.

Other significant changes:

* Improve support for building on Windows platforms. [#1342]

* Fix a bug where WiredTiger could race closing handles. [#1336]

* Enhance performance when hot pages in cache are growing rapidly. [#1317]


src/mongo/client/examples/mongoperf.cpp
/*    Copyright 2009 10gen Inc.
 *
 *    This program is free software: you can redistribute it and/or  modify
 *    it under the terms of the GNU Affero General Public License, version 3,
 *    as published by the Free Software Foundation.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Affero General Public License for more details.
 *
 *    You should have received a copy of the GNU Affero General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 *    As a special exception, the copyright holders give permission to link the
 *    code of portions of this program with the OpenSSL library under certain
 *    conditions as described in each individual source file and distribute
 *    linked combinations including the program with the OpenSSL library. You
 *    must comply with the GNU Affero General Public License in all respects
 *    for all of the code used other than as permitted herein. If you modify
 *    file(s) with this exception, you may extend this exception to your
 *    version of the file(s), but you are not obligated to do so. If you do not
 *    wish to do so, delete this exception statement from your version. If you
 *    delete this exception statement from all source files in the program,
 *    then also delete it in the license file.
 */

/* 
   How to build and run:

   scons mongoperf
   ./mongoperf -h
*/


// note: mongoperf is an internal mongodb utility
// so we define the following macro
#define MONGO_EXPOSE_MACROS 1

#include "mongo/platform/basic.h"

#include <iostream>

#include <boost/filesystem/operations.hpp>
#include <boost/thread/thread.hpp>

#include "mongo/db/jsobj.h"
#include "mongo/db/json.h"
#include "mongo/platform/atomic_word.h"
#include "mongo/util/allocator.h"
#include "mongo/util/logfile.h"
#include "mongo/util/mmap.h"
#include "mongo/util/mongoutils/str.h"
#include "mongo/util/time_support.h"
#include "mongo/util/timer.h"
#include "mongo/util/processinfo.h"


using namespace std;
using namespace mongo;

int dummy;
unsigned recSizeKB;
LogFile *lf = 0;
MemoryMappedFile *mmfFile;
char *mmf = 0;
bo options;
unsigned long long len; // file len
const unsigned PG = 4096;
unsigned nThreadsRunning = 0;

// as this is incremented A LOT, at some point this becomes a bottleneck if very high ops/second (in cache) things are happening.
AtomicUInt32 iops;

SimpleMutex m("mperf");

int syncDelaySecs = 0;

void syncThread() {
    while( 1 ) {
        mongo::Timer t;
        mmfFile->flush(true);
        cout << "                                                     mmf sync took " << t.millis() << "ms" << endl;
        sleepsecs(syncDelaySecs);
    }
}

char* round(char* x) {
    size_t f = (size_t) x;
    char *p = (char *) ((f+PG-1)/PG*PG);
    return p;
}

struct Aligned {
    char x[8192];
    char* addr() { return round(x); }
}; 

unsigned long long rrand() { 
    // RAND_MAX is very small on windows
    return (static_cast<unsigned long long>(rand()) << 15) ^ rand();
}

void workerThread() {
    bool r = options["r"].trueValue();
    bool w = options["w"].trueValue();
    cout << "read:" << r << " write:" << w << endl;
    long long su = options["sleepMicros"].numberLong();
    Aligned a;
    while( 1 ) { 
        unsigned long long rofs = (rrand() * PG) % len;
        unsigned long long wofs = (rrand() * PG) % len;
        const unsigned P = PG/1024;
        if( mmf ) { 
            if( r ) {
                for( unsigned p = P; p <= recSizeKB; p += P ) {
                    if( rofs < len ) 
                        dummy += mmf[rofs];
                    rofs += PG;
                }
                iops.fetchAndAdd(1);
            }
            if( w ) {
                for( unsigned p = P; p <= recSizeKB; p += P ) {
                    if( wofs < len )
                        mmf[wofs] = 3;
                    wofs += PG;
                }
                iops.fetchAndAdd(1);
            }
        }
        else {
            if( r ) {
                lf->readAt(rofs, a.addr(), recSizeKB * 1024);
                iops.fetchAndAdd(1);
            }
            if( w ) {
                lf->writeAt(wofs, a.addr(), recSizeKB * 1024);
                iops.fetchAndAdd(1);
            }
        }
        long long micros = su / nThreadsRunning;
        if( micros ) {
            sleepmicros(micros);
        }
    }
}

void go() {
    verify( options["r"].trueValue() || options["w"].trueValue() );

    recSizeKB = options["recSizeKB"].numberInt();
    if( recSizeKB == 0 )
        recSizeKB = 4;
    verify( recSizeKB <= 64000 && recSizeKB > 0 );

    MemoryMappedFile f;
    cout << "creating test file size:";
    len = options["fileSizeMB"].numberLong();
    if( len == 0 ) len = 1;
    cout << len << "MB ..." << endl;

    if( 0 && len > 2000 && !options["mmf"].trueValue() ) { 
        // todo make tests use 64 bit offsets in their i/o -- i.e. adjust LogFile::writeAt and such
        cout << "\nsizes > 2GB not yet supported with mmf:false" << endl; 
        return;
    }
    len *= 1024 * 1024;
    const char *fname = "./mongoperf__testfile__tmp";
    try {
        boost::filesystem::remove(fname);
    }
    catch(...) { 
        cout << "error deleting file " << fname << endl;
        return;
    }
    lf = new LogFile(fname,true);
    const unsigned sz = 1024 * 1024 * 32; // needs to be big as we are using synchronousAppend.  if we used a regular MongoFile it wouldn't have to be
    char *buf = (char*) mongoMalloc(sz+4096);
    const char *p = round(buf);
    for( unsigned long long i = 0; i < len; i += sz ) { 
        lf->synchronousAppend(p, sz);
        if( i % (1024ULL*1024*1024) == 0 && i ) {
            cout << i / (1024ULL*1024*1024) << "GB..." << endl;
        }
    }
    BSONObj& o = options;

    if( o["mmf"].trueValue() ) { 
        delete lf;
        lf = 0;
        mmfFile = new MemoryMappedFile();
        mmf = (char *) mmfFile->map(fname);
        verify( mmf );

        syncDelaySecs = options["syncDelay"].numberInt();
        if( syncDelaySecs ) {
            boost::thread t(syncThread);
        }
    }

    cout << "testing..."<< endl;

    cout << "options:" << o.toString() << endl;
    unsigned wthr = 1;
    if( !o["nThreads"].eoo() ) {
        wthr = (unsigned) o["nThreads"].Int();
    }
    cout << "wthr " << wthr << endl;

    if( wthr < 1 ) { 
        cout << "bad threads field value" << endl;
        return;
    }
    unsigned i = 0;
    unsigned d = 1;
    unsigned &nthr = nThreadsRunning;
    while( 1 ) {
        if( i++ % 8 == 0 ) {
            if( nthr < wthr ) {
                while( nthr < wthr && nthr < d ) {
                    nthr++;
                    boost::thread w(workerThread);
                }
                cout << "new thread, total running : " << nthr << endl;
                d *= 2;
            }
        }
        sleepsecs(1);
        unsigned long long w = iops.loadRelaxed();
        iops.store(0);
        w /= 1; // 1 secs
        cout << w << " ops/sec ";
        if( mmf == 0 ) 
            // only writing 4 bytes with mmf so we don't say this
            cout << (w * PG / 1024 / 1024) << " MB/sec";
        cout << endl;
    }
}

int main(int argc, char *argv[]) {

    try {
        cout << "mongoperf" << endl;

        if( argc > 1 ) { 
cout <<

"\n"
"usage:\n"
"\n"
"  mongoperf < myjsonconfigfile\n"
"\n"
"  {\n"
"    nThreads:<n>,     // number of threads (default 1)\n"
"    fileSizeMB:<n>,   // test file size (default 1MB)\n"
"    sleepMicros:<n>,  // pause for sleepMicros/nThreads between each operation (default 0)\n"
"    mmf:<bool>,       // if true do i/o's via memory mapped files (default false)\n"
"    r:<bool>,         // do reads (default false)\n"
"    w:<bool>,         // do writes (default false)\n"
"    recSizeKB:<n>,    // size of each write (default 4KB)\n"
"    syncDelay:<n>     // secs between fsyncs, like --syncdelay in mongod. (default 0/never)\n"
"  }\n"
"\n"
"mongoperf is a performance testing tool. the initial tests are of disk subsystem performance; \n"
"  tests of mongos and mongod will be added later.\n"
"most fields are optional.\n"
"non-mmf io is direct io (no caching). use a large file size to test making the heads\n"
"  move significantly and to avoid i/o coalescing\n"
"mmf io uses caching (the file system cache).\n"
"\n"

<< endl;
            return EXIT_SUCCESS;
        }

        cout << "use -h for help" << endl;

        char input[1024];
        memset(input, 0, sizeof(input));
        cin.read(input, 1000);
        if( *input == 0 ) { 
            cout << "error no options found on stdin for mongoperf" << endl;
            return EXIT_FAILURE;
        }

        string s = input;
        mongoutils::str::stripTrailing(s, " \n\r\0x1a");
        try { 
            options = fromjson(s);
        }
        catch(...) { 
            cout << "couldn't parse json options. input was:\n|" << s << "|" << endl;
            return EXIT_FAILURE;
        }
        cout << "parsed options:\n" << options.toString() << endl;
        ProcessInfo::initializeSystemInfo();

        go();
    } 
    catch(DBException& e) { 
        cout << "caught DBException " << e.toString() << endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}

